dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java

```diff
@@ -177,6 +183,19 @@ protected void modified(Map<Object, Object> properties) {
                 defaultProperties.put(key, value);
         }
     }
+
+    @Reference(service = JspVersion.class, cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    protected synchronized void setVersion(ServiceReference<JspVersion> reference) {
+      this.versionRef = reference;
+      JSPExtensionFactory.CURRENT_LOADED_VERSION = (String) reference.getProperty("version");
```

Comment:
Setting this static variable dynamically means that if the static method `getLoadedJspSpecLevel` is called early then it could return `null`.  It seems that `AbstractJSPExtensionServletWrapper` is the only place calling this static.  Is there something passed into the constructor of `AbstractJSPExtensionServletWrapper` that has access to a service component that can be injected with the JspVersion or this JSPExtensionFactory

Reply:
Sorry for the long delay.  I didn't see a way to pass in the variable. Additionally, I did not experience a null values during my testing. 

Since this is quite similar to the servlet version implementation, I think it should be fine for JSP?

Reply:
I think this will work, but still should be considered some amount of technical dept.  If I follow the code correctly the `JSPExtensionFactory` ends up driving the construction of `AbstractJSPExtensionServletWrapper` through inherited method `com.ibm.ws.jsp.webcontainerext.AbstractJSPExtensionProcessor.createServletWrapper(IServletConfig)` when `JSPExtensionFactory` creates the `com.ibm.ws.jsp.webcontainerext.JSPExtensionProcessor` from the method `com.ibm.ws.jsp.webcontainerext.JSPExtensionFactory.createProcessor(IServletContext, JspXmlExtConfig, JspClassloaderContext)`

It seems that somehow the version could be passed down through this code path, but I'm not sure if this is the only usage of `AbstractJSPExtensionServletWrapper`.  It seems there is also an extending class `com.ibm.ws.jsp.inmemory.webcontainerext.AbstractInMemoryJSPExtensionServletWrapper` but I don't see how that class is ever used.

Reply:
As the web app is initialized, JSP page is translated to servlet code, then compiled. When the actual compilation occurs (compile on startup vs compile during a request) depends on some web.xml parameters. 

The translation phase needs a JSP class version associated with it (GeneratorUtilsExtImpl.getClassFileVersion). This version is later used to check if [page was compiled against a different JSP version](https://github.com/volosied/open-liberty/blob/3d311abf451e4b4bdedff40d3f1054dcb23e674e/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/AbstractJSPExtensionServletWrapper.java#L377-L380) (2.2 vs 2.3, for example) in the AbstractJSPExtensionServletWrapper  _checkForTranslation method 

This comparison should be done only during new requests as it’s not needed during startup (translate/compile only occurs once).  The translationRequired flag skips the compilation version check. 

I added a Thread.sleep in the new setVersion method to see if it creates any problems with the versions, but everything still worked properly. (I"m guessing the later Threads don't kick off until the majority of the liberty initializations finish?)

In short, I don’t think there shouldn’t be any issues with AbstractJSPExtensionServletWrapper since when  JSPExtensionFactory.getLoadedPagesSpecLevel is called the pages have compiled at least once and that means there must have been enough time for setVersion to be called. 

In Jsp flow, [GeneratorUtilsExtImpl.getClassFileVersion](https://github.com/volosied/open-liberty/blob/3d311abf451e4b4bdedff40d3f1054dcb23e674e/dev/com.ibm.ws.jsp.2.3/src/com/ibm/ws/jsp23/translator/visitor/generator/GeneratorUtilsExtImpl.java#L95)(which retrieves the version) is called first since translation is one of the earlier steps in the JSP process.  

As for the flow you mentioned in your post the stack trace is below. The createServletWrapper method is called between checkForTranslation & findWrapper.  
```
JSPExtensionServletWrapper(AbstractJSPExtensionServletWrapper).checkForTranslation(HttpServletRequest) line: 245	
JSPExtensionProcessor(AbstractJSPExtensionProcessor).findWrapper(HttpServletRequest, HttpServletResponse) line: 508	
JSPExtensionProcessor(AbstractJSPExtensionProcessor).getServletWrapper(ServletRequest, ServletResponse) line: 351	
WebApp31(WebApp).handleRequest(ServletRequest, ServletResponse, HttpInboundConnection) line: 5027	
DynamicVirtualHost$2.handleRequest(ServletRequest, ServletResponse) line: 314	
WebContainer(WebContainer).handleRequest(IRequest, IResponse, VirtualHost, RequestProcessor) line: 1005	
DynamicVirtualHost$2.run() line: 279	
```